### PR TITLE
Add integration tests for skill templates [skip ci]

### PR DESCRIPTION
**Pull request by rishic3**
Contributes to #15013.

### Description

Adds integration tests for skill templates. (These were originally included in #15058 but were deferred into this separate PR). The tests basically verify that the templates work out of the box: builds work, scripts run, validations catch errors they're supposed to catch, etc. 

The tests create a dummy project, deterministically fill in all of the TODO methods (using the method stubs under `fixtures`) and add UDF implementations (using the complete source files under `resources`), and then run the sequence of scripts that the agent would go through. 

This will be wired into CI in #15140.

### Checklists

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [X] No user-facing change

Testing
- [X] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [ ] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [X] Not required

### CODE REVIEW

File: skills/tests/utils.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    """Run a bash script with the given arguments."""
+    cmd = ["bash", script_path]
+    if args:
+        cmd.extend(args)
+    result = subprocess.run(
+        cmd,
+        capture_output=True,
+        text=True,
+        timeout=timeout,
+    )
```
[the remark this answers is not in the document]
Comment (rishic3) [reply skills/tests/utils.py:61]: Navigation to the right directory is handled by the bash script.

File: skills/tests/utils.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def run_mvn(
+    work_dir: str,
+    *goals: str,
+    extra_args: Optional[list[str]] = None,
+    timeout: int = 300,
+) -> subprocess.CompletedProcess:
+    """Run Maven in work_dir with the given goals."""
+    cmd = ["mvn", *goals, "-q"]
```
[the remark this answers is not in the document]
Comment (rishic3) [reply skills/tests/utils.py:33]: I think -q just lowers the log level, we'll never lose errors. Also the bench scripts aren't run with surefire.

File: skills/docs/dev/TESTING.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+Run the integration tests:
+
+```bash
+pytest -m slow -s
+```
+
+These tests deterministically fill in the template project from `skills/udf-gen-test/templates/` with fixture implementations, then actually compile and run Spark tests and benchmark scripts locally.
+
+Thus they require JDK, Maven and Maven repository access, and a GPU environment.
```
[the remark this answers is not in the document]
Comment (rishic3) [reply skills/docs/dev/TESTING.md]: Clarified.

File: skills/tests/fixtures.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+"""
+Source fixtures for the JVM template integration tests.
+"""
+
+from pathlib import Path
+
+
+def _read_resource(name: str) -> str:
+    return (Path(__file__).parent / "resources" / name).read_text()
```
[the remark this answers is not in the document]
Comment (rishic3) [reply skills/tests/fixtures.py]: Done

File: skills/tests/test_jvm_templates.py [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+import stat
+import tempfile
+from pathlib import Path
+
+import pytest
+
+from . import fixtures as fx
+from .utils import replace_scala_todo_method, run_mvn, run_script
+
+pytestmark = pytest.mark.slow
```
[the remark this answers is not in the document]
Comment (rishic3) [reply skills/tests/test_jvm_templates.py:19]: I don't think we should silently skip. The CI image will definitely have maven.

File: skills/tests/test_jvm_templates.py [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+                "-Dexec.classpathScope=compile",
+                "-Dexec.args=--rows 1000 --validate --spark-conf spark.master=local[*]",
+            ],
+        )
+        assert (
+            result.returncode != 0
+        ), "GenData --validate should have failed with wrong schema"
+        assert (
+            "org.apache.spark.sql.AnalysisException" in result.stderr
+        ), "Expected to see schema error message"
```
[the remark this answers is not in the document]
Comment (rishic3) [reply skills/tests/test_jvm_templates.py:527]: This is a duplicate of https://github.com/NVIDIA/cudf-spark/pull/15131#discussion_r3463597891

**Comment by rishic3** [comment]
build

**Comment by rishic3** [comment]
build

**Comment by rishic3** [comment]
build

**Review by wjxiz1992 — DISMISSED** [review]

File: skills/docs/dev/TESTING.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+source .venv/bin/activate
+pip install -e ".[dev]"
+```
+
+## Fast Tests
+
+Run the fast tests:
+
+```bash
+pytest -m "not slow"
```
Comment (sayedbilalbari) [inline skills/docs/dev/TESTING.md:20]: qq: What's the trigger plan for these tests ? Will they run as part of pre-merge always ?

File: skills/pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,38 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+[build-system]
+requires = ["setuptools>=61.0"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "aether-agent"
```
Comment (sayedbilalbari) [inline skills/pyproject.toml:9]: qq: The name aether is still being used here ?

File: skills/tests/utils.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    return result
+
+
+def replace_scala_todo_method(source: str, method_name: str, new_body: str) -> str:
+    """
+    Replace a TODO method stub in a Scala source file with a real implementation.
+    Assumes stubs look like "def foo(...) = ???"
+    """
+    pattern = re.compile(
+        r"  def " + re.escape(method_name) + r"\b.*?\?\?\?",
```
Comment (sayedbilalbari) [inline skills/tests/utils.py]: Since the regex follows this pattern -> def first(): <any characters> ??? . A scenario like this -
```
def first(): Unit = {
}
def second(): Unit = ???
```
Searching for match for first would still pass. Maybe constraint it

File: skills/tests/utils.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+        sys.stdout.write(_indent(result.stdout))
+    if result.stderr:
+        sys.stderr.write(_indent(result.stderr))
+
+
+def run_mvn(
+    work_dir: str,
+    *goals: str,
+    extra_args: Optional[list[str]] = None,
+    timeout: int = 300,
```
Comment (sayedbilalbari) [inline skills/tests/utils.py]: Have we always seen this default timeout of 5 minutes to work ? Do we know cold-cache behavior when building for the integration tests ?

File: skills/tests/fixtures.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+BENCH_GENERATE = """\
+  def generateSyntheticData(
+      spark: SparkSession,
+      numRows: Long,
+      numPartitions: Int
+  ): DataFrame = {
+    val baseDF = spark.range(0, numRows, 1, numPartitions)
+    baseDF.select(
+      col("id"),
+      (rand() * 1000).cast(IntegerType).alias("value")
```
Comment (sayedbilalbari) [inline skills/tests/fixtures.py]: nit: Would we want to seed the rand() to make sure it always generates the same set of values ( makes this repeatable )

File: skills/tests/fixtures.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+# ---------------------------------------------------------------------------
+
+CREATE_TEST_DATA = """\
+  def createTestData(spark: SparkSession): DataFrame = {
+    val schema = StructType(Seq(
+      StructField("id", IntegerType, nullable = false),
+      StructField("value", IntegerType, nullable = true)
+    ))
+    val testData = Seq(
+      Row(1, 123),
```
Comment (sayedbilalbari) [inline skills/tests/fixtures.py:44]: qq: Is this dataset enough to cover all cases ? What about the boundary cases - INT_MAX etc. ? Is that even needed ?

File: skills/tests/test_frontmatter.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+from pathlib import Path
+
+import pytest
+import yaml
+
+SKILLS_DIR = Path(__file__).resolve().parents[1]
+SKILL_FILES = sorted(SKILLS_DIR.glob("*/SKILL.md"))
+
+
+def _read_frontmatter(path: Path) -> str:
```
Comment (sayedbilalbari) [inline skills/tests/test_frontmatter.py:17]: qq: Would we want to expand this to also check non-empty values for name, description ?

File: skills/tests/test_jvm_templates.py [hunk 14]
```diff
@@ -0,0 +1,522 @@
...
+        )
+        assert result.returncode == 0, f"{suite} failed"
+
+
+class TestBench:
+    """Test the benchmark pipeline (GenData + BenchRunner)."""
+
+    def test_validate(self, project_with_fixtures):
+        """GenData validation: generate a small dataset and run validation."""
+        proj, _, target = project_with_fixtures
```
Comment (sayedbilalbari) [inline skills/tests/test_jvm_templates.py:382]: Codex pointed out that the documented flow to generate data is through -
```
./run_gen_data.sh --rows 1000 --validate
```
Would we want to update this to use that flow ?

File: skills/tests/test_jvm_templates.py [hunk 15]
```diff
@@ -0,0 +1,522 @@
...
+@pytest.fixture(scope="class", params=TARGETS, ids=lambda t: t)
+def project_with_broken_gpu(request):
+    """Project with deliberately broken GPU implementation."""
+    target = request.param
+    lang = "scala"
+    tmp_dir, proj = _build_project_dir()
+
+    def _break_gpu_source(source: str) -> str:
+        # Change multiplier from 2 to 3
+        return source.replace("fromInt(2)", "fromInt(3)").replace("* 2", "* 3")
```
Comment (sayedbilalbari) [inline skills/tests/test_jvm_templates.py]: Codex pointed out -
For the scala cuDF fixture, this modifies:
- The GPU evaluateColumnar implementation
- The CPU fallback apply implementation

the comparison fails even if Spark unexpectedly executes the CPU fallback. It does not prove that the broken GPU implementation was reached

File: skills/tests/resources/IntegerMultiplyBy2NativeRapidsUDF.java [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package com.udf;
+
+import ai.rapids.cudf.ColumnVector;
+import com.nvidia.spark.RapidsUDF;
+import org.apache.hadoop.hive.ql.exec.UDF;
+import org.apache.spark.sql.api.java.UDF1;
+
+public class IntegerMultiplyBy2NativeRapidsUDF extends UDF
```
Comment (sayedbilalbari) [inline skills/tests/resources/IntegerMultiplyBy2NativeRapidsUDF.java]: nit: Codex pointed out -
The CPU reference is a Java UDF1: it does not extend Hive's UDF. Therefore this wrapper does not need - extends UDF or the corresponding import

**Review by sayedbilalbari — COMMENTED** [review]
Thanks @rishic3 , overall looks good. Had some nits and questions

File: skills/docs/dev/TESTING.md (hunk 7 above)
Comment (rishic3) [reply skills/docs/dev/TESTING.md:20]: Yup, from https://github.com/NVIDIA/cudf-spark/issues/14977#issuecomment-4688030833:

> This could be a separate stage/step in pre-merge CI that runs when changes under skills/* are detected. At the same time, if only skills/* is touched, we can skip the regular stages/steps.

File: skills/pyproject.toml (hunk 8 above)
Comment (rishic3) [reply skills/pyproject.toml:9]: Yes that's intentional, still the tentative name as of now.

File: skills/tests/utils.py (hunk 9 above)
Comment (rishic3) [reply skills/tests/utils.py]: Done

File: skills/tests/utils.py (hunk 10 above)
Comment (rishic3) [reply skills/tests/utils.py]: Works fine locally on cold cache, but i bumped to 10 min anyway to be safe.

File: skills/tests/fixtures.py (hunk 11 above)
Comment (rishic3) [reply skills/tests/fixtures.py]: Can't hurt. Done.

File: skills/tests/fixtures.py (hunk 12 above)
Comment (rishic3) [reply skills/tests/fixtures.py:44]: I think it's fine; the purpose is just to exercise the harness, not stress test the UDF.

File: skills/tests/test_frontmatter.py (hunk 13 above)
Comment (rishic3) [reply skills/tests/test_frontmatter.py:17]: Sure, done.

File: skills/tests/test_jvm_templates.py (hunk 14 above)
Comment (rishic3) [reply skills/tests/test_jvm_templates.py:382]: The bash scripts are covered in `test_spark_e2e` so I think we're good. It's nice to also cover invoking directly here.

File: skills/tests/test_jvm_templates.py (hunk 15 above)
Comment (rishic3) [reply skills/tests/test_jvm_templates.py]: Good point. I addressed this (and refactored so the handling is more explicit).

File: skills/tests/resources/IntegerMultiplyBy2NativeRapidsUDF.java (hunk 16 above)
Comment (rishic3) [reply skills/tests/resources/IntegerMultiplyBy2NativeRapidsUDF.java]: Good catch. Fixed.

File: skills/tests/fixtures.py [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+    cls=CPU_UDF_NAME,
+    invoke="call",
+)
+
+
+_MICRO_EXECUTE_GPU_METHOD = """\
+  def executeGpu(table: Table, numRows: Int): Unit = {{
+    val udf = new com.udf.{cls}()
+    withResource(udf.evaluateColumnar(numRows, table.getColumn(1))) {{ _ => }}
+  }}"""
```
[the remark this answers is not in the document]
Comment (rishic3) [reply skills/tests/fixtures.py:197]: This was changed in https://github.com/NVIDIA/cudf-spark/pull/15116. I think you have the wrong main branch.

**Comment by rishic3** [comment]
build

**Review by sayedbilalbari — APPROVED** [review]
Thanks @rishic3 , LGTM !
